wgpu-core/src/command/mod.rs

```diff
@@ -422,15 +434,75 @@ impl<A: HalApi> CommandBuffer<A> {
     ) -> Result<Arc<Self>, CommandEncoderError> {
         let storage = hub.command_buffers.read();
         match storage.get(id.into_command_buffer_id()) {
-            Ok(cmd_buf) => match cmd_buf.data.lock().as_ref().unwrap().status {
-                CommandEncoderStatus::Recording => Ok(cmd_buf.clone()),
-                CommandEncoderStatus::Finished => Err(CommandEncoderError::NotRecording),
-                CommandEncoderStatus::Error => Err(CommandEncoderError::Invalid),
-            },
+            Ok(cmd_buf) => {
+                let mut cmd_buf_data = cmd_buf.data.lock();
+                let cmd_buf_data = cmd_buf_data.as_mut().unwrap();
+                match cmd_buf_data.status {
+                    CommandEncoderStatus::Recording => Ok(cmd_buf.clone()),
+                    CommandEncoderStatus::Locked => {
+                        // Any operation on a locked encoder is required to put it into the invalid/error state.
+                        // See https://www.w3.org/TR/webgpu/#encoder-state-locked
+                        cmd_buf_data.encoder.discard();
+                        cmd_buf_data.status = CommandEncoderStatus::Error;
+                        Err(CommandEncoderError::Locked)
+                    }
+                    CommandEncoderStatus::Finished => Err(CommandEncoderError::NotRecording),
+                    CommandEncoderStatus::Error => Err(CommandEncoderError::Invalid),
+                }
+            }
             Err(_) => Err(CommandEncoderError::Invalid),
         }
     }
 
+    /// Return the [`CommandBuffer`] for `id` and if successful puts it into the [`CommandEncoderStatus::Locked`] state.
+    ///
+    /// See [`CommandBuffer::get_encoder`].
+    /// Call [`CommandBuffer::unlock_encoder`] to put the [`CommandBuffer`] back into the [`CommandEncoderStatus::Recording`] state.
+    fn lock_encoder(
```

Comment:
For code reuse purposes, could this be a call to get_encoder, then locking it? The bodies are basically identical.

Reply:
yeah that would work, but then I'd need to take the lock again. So this would like like:
```rust
    fn lock_encoder(
        hub: &Hub<A>,
        id: id::CommandEncoderId,
    ) -> Result<Arc<Self>, CommandEncoderError> {
        Self::get_encoder(hub, id).map(|result| {
            result.data.lock().as_mut().unwrap().status = CommandEncoderStatus::Locked;
            result
        })
    }
```
this works & passes tests but not quite happy with it :/

Reply:
decided to solve it with
```rust
    fn get_encoder_impl(
        hub: &Hub<A>,
        id: id::CommandEncoderId,
        lock_on_acquire: bool,
    ) -> Result<Arc<Self>, CommandEncoderError> {
```